### ISSUE TITLE
Send multiple grpc requests at the same time, when one of them fails, then other requests all call back "Transport became inactive".

### ISSUE THREAD

**Issue by jackystd** [grpc/grpc-swift#1421]
```
        let (host, port) = address.convertAddress()
        let useTLS = (port == 443)
        let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
        let builder = useTLS ? ClientConnection.usingTLSBackedByNIOSSL(on: group) : ClientConnection.insecure(group: group)
        let channel = builder.connect(host: host, port: port)
        channel.connectivity.delegate = self
```

We upgraded from grpc0.x  to grpc1.7. When the v0.x sends multiple requests at the same time, it will not affect each other when an error occurs, and the v1.7 will call back all 'Transport became inactive' after the error occurs.

Is this normal? I expect multiple requests to not affect each other.

**Comment by glbrntt** [comment]
It depends why they fail. If the connection has failed then it's highly likely that all in flight RPCs will fail.

For what it's worth, you should prefer using the `GRPCChannelPool` over `ClientConnection`. It is a little more resilient to failures.

**Comment by jackystd** [comment]
> It depends why they fail. If the connection has failed then it's highly likely that all in flight RPCs will fail.
> 
> For what it's worth, you should prefer using the `GRPCChannelPool` over `ClientConnection`. It is a little more resilient to failures.


Specifically, the error we have is a 502, can I customize the error handling?

**Comment by glbrntt** [comment]
That doesn't seem right: a 502 will be mapped to a status unavailable message and should only fail the RPC that receives the 502. 

Are you sure that the 502 is the cause of the connection being closed? Do you have some logs we could look at?

**Comment by jackystd** [comment]
> That doesn't seem right: a 502 will be mapped to a status unavailable message and should only fail the RPC that receives the 502.
> 
> Are you sure that the 502 is the cause of the connection being closed? Do you have some logs we could look at?


When I receive the first response with a status code of 502, the ClientChannel will close the connection in the function: **handleError()**, and subsequent requests will fail(return with 'Transport became inactive') until the reconnection succeeds.

See the file: ClientTransport.swift(line:399).
```
 /// On-loop implementation of `handleError(_:)`.
  private func _handleError(_ error: Error) {
    self.callEventLoop.assertInEventLoop()

    switch self.state.handleError() {
    case .doNothing:
      ()

    case .propagateError:
      self.forwardErrorToInterceptors(error)
      self.failBufferedWrites(with: error)

    case .propagateErrorAndClose:
      self.forwardErrorToInterceptors(error)
      self.failBufferedWrites(with: error)
      **self.channel?.close(mode: .all, promise: nil)**
    }
  }
```

**Comment by glbrntt** [comment]
`self.channel` is the channel of the HTTP/2 stream. Its parent channel is for the HTTP/2 connection which should is should not be closed as a result of the stream channel closing.

If something is closing the connection channel, I don't think it's this line of code. Do you have some logs we could look at?

**Comment by jackystd** [comment]
> look at?

Looking forward to your reply

**Comment by glbrntt** [comment]
Do you have trace logs? (You can set `logger.logLevel = .trace`)

**Comment by Banck** [comment]
Seems it's the same issue, which I had before: https://github.com/grpc/grpc-swift/issues/1341
We resolved it for now by creating new ClientConnection for every new requests

**Comment by jackystd** [comment]
> Seems it's the same issue, which I had before: #1341 We resolved it for now by creating new ClientConnection for every new requests

Thanks, I've tried this and it works, but I don't think it's best solution.
Specifically, I found that the time per request increased significantly.

**Comment by glbrntt** [comment]
Thanks for providing logs. It looks like metadata is missing from them though; could you provide them with metadata?

**Comment by glbrntt** [comment]
No, I mean metadata attached to logs. The log formatter being used here is dropping metadata (which includes things like the request ID, connection ID, etc.)

**Comment by jackystd** [comment]
> No, I mean metadata attached to logs. The log formatter being used here is dropping metadata (which includes things like the request ID, connection ID, etc.)


I uploaded the complete log to [https://github.com/jackystd/logs/blob/main/grpc_disconnect.log](url) , please check if metadata is included.

**Comment by pacu** [comment]
> It depends why they fail. If the connection has failed then it's highly likely that all in flight RPCs will fail.
> 
> For what it's worth, you should prefer using the `GRPCChannelPool` over `ClientConnection`. It is a little more resilient to failures.

@glbrntt I'm trying this out related to the issue #1273  which has been really helpful. and the recurring errors seem to be gone. 

My question is. How can I hook a connection delegate to the channel as I had with the ClientConnection?

**Comment by glbrntt** [comment]
There is no connection delegate for the pool.
